src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java

```diff
@@ -156,31 +176,61 @@ private static void writeReadsSingle(
+        }
+        Ordering<K> ordering = Ordering$.MODULE$.comparatorToOrdering(comparator);
+        ClassTag<K> tag = ClassTag$.MODULE$.apply(keyClass);
+        RangePartitioner<K, V> partitioner = new RangePartitioner<>(numPartitions, rdd.rdd(), true, ordering, tag);
+        return rdd.repartitionAndSortWithinPartitions(partitioner, comparator);
     }
 
     private static void deleteHadoopFile(String fileToObliterate) throws IOException {
         Configuration conf = new Configuration();
         FileSystem fs = FileSystem.get(conf);
-        fs.delete(new Path(fileToObliterate),true);
+        fs.delete(new Path(fileToObliterate), true);
     }
 
-    private static void renameHadoopSingleShard(String outputFile) throws IOException {
-        // At this point, the file is in outputFile/part-r-0000. We move it out of that dir and into it's own well-named
-        // file.  This will not work if the last step was a merge (assuming saveAsNewAPIHadoopFile doesn't cause the
-        // last step to be a reduce).
+    private static void mergeBam(String outputFile, SAMFileHeader header) throws IOException {
+        // At this point, the part files (part-r-00000, part-r-00001, etc) are in a directory named outputFile.
+        // We concatenate them into a file named outputFile.
+
         Configuration conf = new Configuration();
         FileSystem fs = FileSystem.get(conf);
-        String ouputParentDir = outputFile.substring(0, outputFile.lastIndexOf("/") + 1);
+        String outputParentDir = outputFile.substring(0, outputFile.lastIndexOf("/") + 1);
         // First, check for the _SUCCESS file.
         String successFile = outputFile + "/_SUCCESS";
         if (!fs.exists(new Path(successFile))) {
             throw new GATKException("unable to find " + successFile + " file");
         }
-        String tmpPath = ouputParentDir + "tmp" + System.currentTimeMillis();
-        fs.rename(new Path(outputFile + "/part-r-00000"), new Path(tmpPath));
+        String tmpPath = outputParentDir + "tmp" + System.currentTimeMillis();
+        fs.rename(new Path(outputFile), new Path(tmpPath));
         fs.delete(new Path(outputFile), true);
-        fs.rename(new Path(tmpPath), new Path(outputFile));
+
+        try (final OutputStream out = fs.create(new Path(outputFile))) {
+            new SAMOutputPreparer().prepareForRecords(out, SAMFormat.BAM, header); // write the header
+            mergeInto(out, new Path(tmpPath), conf);
+            out.write(BlockCompressedStreamConstants.EMPTY_GZIP_BLOCK); // add the BGZF terminator
+        }
+    }
+
+    private static void mergeInto(OutputStream out, Path directory, Configuration conf) throws IOException {
+        final FileSystem fs = directory.getFileSystem(conf);
+        final FileStatus[] parts = fs.globStatus(new Path(directory, "part-r-[0-9][0-9][0-9][0-9][0-9]*"));
+        for (final FileStatus part : parts) {
+            try (final InputStream in = fs.open(part.getPath())) {
```

Comment:
From @davidadamsphd (github lost this comment during rebase):

It's interesting that you can just concat the gzipped files together, I would have thought that the end of file marking on the first BAM would trip up the combined bam (think it's done early). Do you know why that's not an issue? It looks like `readsSinkTest` would have caught this if it was an issue, right?

Also, this have the side effect of being ever-so-slightly less efficient than writing it as one BAM because the last block of each file isn't completely full.

I think both of these are worth documenting.